src/options.c

```diff
@@ -484,27 +481,38 @@ void optionsParse(int argc, char *argv[])
             break;
         }
     }
-
-    /* Now the leftovers, which must be files */
-    while (optind < argc) {
-        /* If recursive is NOT set, but the only argument is a directory
-           name, we grab all the files in there, but not subdirs */
-        if (!opt.outputFile) {
-            optionsParseFileName(argv[optind++]);
-
-            const bool redirectChar = ( opt.outputFile[0] == '-'
-                                        && opt.outputFile[1] == '\0');
-            if (redirectChar) {
-                optionsParseFileName(getPathOfStdout());
-                opt.overwrite = 1;
-                opt.thumb = THUMB_DISABLED;
-            }
-        } else
-            warnx("unrecognised option %s", argv[optind++]);
+    argv += optind;
+
+    if (*argv) {
+        if (opt.outputFile) {
+            errx(EXIT_FAILURE,
+                "both the --file option and the FILE parameter were used");
+        } else {
+            opt.outputFile = *argv;
+            argv++;
+        }
     }
-
-    /* So that we can safely be called again */
-    optind = 1;
+    if (*argv) {
+        /* If *argv is still not NULL, we have an excess of arguments, so print
+         * out a nice diagnostic message and error exit.
+         */
+        fprintf(stderr, "%s: extraneous non-option arguments: ", getprogname());
+        for (;;) {
+            fputs(*argv, stderr);
+            if (*++argv)
+                fputc(' ', stderr);
+            else
+                break;
+        }
+        fputc('\n', stderr);
+        exit(EXIT_FAILURE);
+    }
+    if (strcmp(opt.outputFile, "-") == 0) {
+        opt.overwrite = 1;
+        opt.thumb = THUMB_DISABLED;
+        opt.outputFile = (/* const */ char *)getPathOfStdout();
+    }
+    opt.outputFile = estrdup(opt.outputFile);
```

Comment:
I noticed that while working on it. There's no reason, actually. `opt.outputFile` gets passed around a little, and is only ever read but not written. It's not `free()`d either. 

I decided to keep things as they are because in the end, it's used by `imPrintf()`, and that function needs some fixing. It outputs to a 4KiB buffer on the stack and then duplicates it. While making this PR, I thought it would be best if `imPrintf()` was modified to assume its `str` parameter was alloc storage duration and grow it as needed, but I just realized a better design would be if the `strdup()` happened as late as possible which is inside `imPrintf()`.

Reply:
If there's no need to dup, then I'd drop the `estrdup` and make `outputFile` const.